solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/CollectionPropertiesZkStateReader.java

```diff
@@ -300,10 +313,11 @@ private VersionedCollectionProps fetchCollectionProperties(String collection, Wa
       throws KeeperException, InterruptedException {
     final String znodePath = getCollectionPropsPath(collection);
     // lazy init cache cleaner once we know someone is using collection properties.
-    if (collectionPropsCacheCleaner == null) {
-      synchronized (zkStateReader.getUpdateLock()) { // Double-checked locking
-        if (collectionPropsCacheCleaner == null) {
-          collectionPropsCacheCleaner = notifications.submit(new CacheCleaner());
+    if (cacheCleanerThread == null) {
+      synchronized (this) {
+        if (cacheCleanerThread == null) {
```

Comment:
Probably should check for `!closed` as well so that we don't start this thread after closing.

Come to think of it, the CacheCleaner ends on interruption and it doesn't looked at "closed" at all.  That's rather brittle; it should loop while `!closed` and not "break" if Interrupted -- this is the typical approach.

Come to think of it further... hey, that CacheCleaner looks like a "scheduled" job every minute.  Could use a ScheduledExecutor and look more "obvious".  That's what I recommend.  I know that changes things here :-)

Reply:
Can we instead revert to the logic 2 commits prior that used the ExecutorService and the Atomic Boolean to ensure the `cacheCleaner` is initialized only once?

Reply:
If you want.
But I think the best solution reflecting the pattern/need for this cache cleaner is a ScheduledExecutorService and no extra state var.  all ExecutorService's *have* state or could use null.

Reply:
I tried this. Using the following code, where cacheCleanerExecutor is defined as 
```
 private final ScheduledExecutorService cacheCleanerExecutor =
      Executors.newSingleThreadScheduledExecutor(new SolrNamedThreadFactory("cacheCleaner"));

```:
and then used in `fetchCollectionProperties()` as follows:
```
 if (cacheCleanerExecutor.isShutdown()) {
      synchronized (this) {
        if (cacheCleanerExecutor.isShutdown()) {
          cacheCleanerExecutor.scheduleAtFixedRate(new CacheCleaner(), 0, 1, TimeUnit.MINUTES);
        }
      }
  }
```

I ran a Junit test and set a breakpoint at `cacheCleanerExecutor.scheduleAtFixedRate`, it is never hit. 

But when I changed to using a state variable like the following, I see the breakpoint being hit to schedule the cacheCleaner thread
```
if (collectionPropsCacheCleanerInitialized.compareAndSet(false, true)) {
      synchronized (this) {
          cacheCleanerExecutor.scheduleAtFixedRate(new CacheCleaner(), 0, 1, TimeUnit.MINUTES);
      }
 }
```

Reply:
fetchCollectionProperties: need to know if there is a task in the executor or not.  If we use the ScheduledThreadPoolExecutor type, then we have methods like `getTaskCount` to know if something is scheduled.
An alternative (maybe simpler?) is to use Timer.
Wether the ScheduledExecutor or Timer, could also use a null instance initially (not final, of course) to differentiate not-started from started which is totally fine too.

Reply:
Ok using ScheduledThreadPoolExecutor works. I updated the PR.

Reply:
Looks good!